### DISCUSSION TITLE
Added custom field of String type in skywalking-browser-receiver-plugin

### DISCUSSION THREAD

**Discussion by PiyushSharma2812** [apache/skywalking#12246] [General]
Hi Team,

I am using skywalking version 9.7 in that i am trying to add field of String type and made the required changes in its dependent files.
And create a **OAL Script** as
**browser_app_page_browser_name = from(BrowserAppPagePerf.browserName).filter(tags["browsername"] != null).map(tags["browsername"])
browser_app_page_browser_name = from(BrowserAppPagePerf.browserName);**
 but all the available methods are integer type is there any way to show the string, we tried map as well  and without providing anything also but getting error. 
 
 Does anyone have any suggestions? Your help would be appreciated.

**Comment by wu-sheng** [comment]
The metric is an abstract concept of the value, labeled value or single number value. It can't be a string for sure.

**Reply by AjinThankachan** [reply]
Thanks for your prompt response.But if we need to add a string something like the service name. Say for example we are getting the browsername and system OS name from client js by changing or adding a new parameter which is relevant in monitoring  and if we need to pass it to skywalking what we thought is create a variable of string type in the browser receiver code and post/push the  value  like browserName: 'Chrome'. So using the OAL it will take the browserName parameter. Is there any possible way to do it ? As of now we appended to BrowserAppPagePerf this parameter as well.

**Reply by wu-sheng** [reply]
You could add anything you like as a parameter. But, you are doing using the new string time in the `from`, which means, its value is built as the metric value. This is wrong.

**Reply by AjinThankachan** [reply]
ok thanks. So can you help me to understand how can i do this? Is there any doc or something is available to add custom parameter as string but not using the from and not as metrics.  Once i am done with the POC i can share the proposal as well. SO please suggest on that context if possible.

**Reply by wu-sheng** [reply]
To be honest, I can't help much as this is your private fork. Meanwhile, I don't know anything has to be changed to add a string field. This should be straightforward from my understanding.
e.g. `org.apache.skywalking.oap.server.core.source.Service` has many string fields.

**Comment by PiyushSharma2812** [comment]
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.skywalking.oap.server.receiver.browser.provider.parser.performance.listener;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppPagePerf;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppPageTraffic;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppPerf;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppPerfSource;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppSingleVersionPerf;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppSingleVersionTraffic;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppTraffic;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppTrafficCategory;
import org.apache.skywalking.oap.server.core.browser.source.BrowserAppTrafficSource;
import org.apache.skywalking.oap.server.core.config.NamingControl;
import org.apache.skywalking.oap.server.core.source.Source;

/**
 * Browser traffic and performance related source.
 */
@RequiredArgsConstructor
class SourceBuilder {
    private final NamingControl namingControl;

    @Getter
    private String service;

    public void setService(final String service) {
        this.service = namingControl.formatServiceName(service);
    }

    @Getter
    private String serviceVersion;

    public void setServiceVersion(final String serviceVersion) {
        this.serviceVersion = namingControl.formatInstanceName(serviceVersion);
    }

    @Getter
    private String patePath;

    public void setPatePath(final String patePath) {
        this.patePath = namingControl.formatEndpointName(service, patePath);
    }

    @Setter
    @Getter
    private long timeBucket;

    // performance data detail
    @Setter
    @Getter
    private int redirectTime;
    @Setter
    @Getter
    private int dnsTime;
    @Setter
    @Getter
    private int ttfbTime;
    @Setter
    @Getter
    private int tcpTime;
    @Setter
    @Getter
    private int transTime;
    @Setter
    @Getter
    private int domAnalysisTime;
    @Setter
    @Getter
    private int fptTime;
    @Setter
    @Getter
    private int domReadyTime;
    @Setter
    @Getter
    private int loadPageTime;
    @Setter
    @Getter
    private int resTime;
    @Setter
    @Getter
    private int sslTime;
    @Setter
    @Getter
    private int ttlTime;
    @Setter
    @Getter
    private int firstPackTime;
    @Setter
    @Getter
    private int fmpTime;
    **@Setter
    @Getter
    private String browserName;**

    private void toSource(Source source) {
        source.setTimeBucket(timeBucket);
    }

    private void toBrowserAppTrafficSource(BrowserAppTrafficSource source) {
        toSource(source);
        source.setTrafficCategory(BrowserAppTrafficCategory.NORMAL);
    }

    /**
     * Browser service meta and traffic metrics related source.
     */
    BrowserAppTraffic toBrowserAppTraffic() {
        BrowserAppTraffic traffic = new BrowserAppTraffic();
        traffic.setName(service);
        toBrowserAppTrafficSource(traffic);
        return traffic;
    }

    /**
     * Browser single version meta and traffic metrics related source.
     */
    BrowserAppSingleVersionTraffic toBrowserAppSingleVersionTraffic() {
        BrowserAppSingleVersionTraffic traffic = new BrowserAppSingleVersionTraffic();
        traffic.setName(serviceVersion);
        traffic.setServiceName(service);
        toBrowserAppTrafficSource(traffic);
        return traffic;
    }

    /**
     * Browser page meta and traffic metrics related source.
     */
    BrowserAppPageTraffic toBrowserAppPageTraffic() {
        BrowserAppPageTraffic traffic = new BrowserAppPageTraffic();
        traffic.setName(patePath);
        traffic.setServiceName(service);
        toBrowserAppTrafficSource(traffic);
        return traffic;
    }

    private void toBrowserAppPerfSource(BrowserAppPerfSource source) {
        toSource(source);
        source.setRedirectTime(redirectTime);
        source.setDnsTime(dnsTime);
        source.setTtfbTime(ttfbTime);
        source.setTcpTime(tcpTime);
        source.setTransTime(transTime);
        source.setDomAnalysisTime(domAnalysisTime);
        source.setFptTime(fptTime);
        source.setDomReadyTime(domReadyTime);
        source.setLoadPageTime(loadPageTime);
        source.setResTime(resTime);
        source.setSslTime(sslTime);
        source.setTtlTime(ttlTime);
        source.setFirstPackTime(firstPackTime);
        source.setFmpTime(fmpTime);
        **source.setBrowserName(browserName);**
    }

    /**
     * Browser service performance related source.
     */
    BrowserAppPerf toBrowserAppPerf() {
        BrowserAppPerf perf = new BrowserAppPerf();
        perf.setName(service);
        toBrowserAppPerfSource(perf);
        return perf;
    }

    /**
     * Browser single version performance related source.
     */
    BrowserAppSingleVersionPerf toBrowserAppSingleVersionPerf() {
        BrowserAppSingleVersionPerf perf = new BrowserAppSingleVersionPerf();
        perf.setName(serviceVersion);
        perf.setServiceName(service);
        toBrowserAppPerfSource(perf);
        return perf;
    }

    /**
     * Browser page performance related source.
     */
    BrowserAppPagePerf toBrowserAppPagePerf() {
        BrowserAppPagePerf perf = new BrowserAppPagePerf();
        perf.setName(patePath);
        perf.setServiceName(service);
        toBrowserAppPerfSource(perf);
        return perf;
    }
}

I made these changes that is highlighted above. Is it possible to create OAL for this?
